optaweb-employee-rostering-backend/src/main/java/org/optaweb/employeerostering/service/roster/RosterGenerator.java

```diff
@@ -668,48 +669,51 @@ public RosterState createRosterState(GeneratorType generatorType, Tenant tenant,
                                        RosterConstraintConfiguration rosterConstraintConfiguration,
                                        RosterState rosterState,
                                        List<Spot> spotList,
-                                       List<ShiftTemplate> shiftTemplateList) {
+                                       List<TimeBucket> timeBucketList) {
         ZoneId zoneId = rosterState.getTimeZone();
         int rotationLength = rosterState.getRotationLength();
-        LocalDate date = rosterState.getLastHistoricDate().with(TemporalAdjusters.previousOrSame(DayOfWeek.MONDAY));
+        LocalDate nextDate = rosterState.getLastHistoricDate().with(TemporalAdjusters.previousOrSame(DayOfWeek.MONDAY));
         LocalDate firstDraftDate = rosterState.getFirstDraftDate();
         LocalDate firstUnplannedDate = rosterState.getFirstUnplannedDate();
 
         List<Shift> shiftList = new ArrayList<>();
-        Map<Pair<Integer, Spot>, List<ShiftTemplate>> dayOffsetAndSpotToShiftTemplateListMap = shiftTemplateList
-                .stream()
-                .collect(groupingBy(shiftTemplate -> Pair.of(shiftTemplate.getStartDayOffset(),
-                                                             shiftTemplate.getSpot())));
-        int dayOffset = 0;
-        while (date.compareTo(firstUnplannedDate) < 0) {
+
+        int nextDayOffset = 0;
+        while (nextDate.compareTo(firstUnplannedDate) < 0) {
```

Comment:
Another loop that is somewhat hard to read.
Breaking it down into smaller parts would help.